File: minute_cCodeGenerator.pl
#!/usr/bin/perl
#-------------------------------------------------------------------------------------------------------------------------------
#                                                                                                       
#                            _/      _/  _/                        _/    _/          _/_/_/_/_/         
#                           _/_/  _/_/      _/_/_/                _/    _/              _/      _/_/    
#                          _/  _/  _/  _/  _/    _/  _/_/_/_/_/  _/    _/  _/_/_/_/_/  _/    _/_/_/_/   
#                         _/      _/  _/  _/    _/              _/    _/              _/    _/          
#                        _/      _/  _/  _/    _/                _/_/                _/      _/_/_/     
#                                                                                                       
#                                                                                                       
#                                                    Minimal Unit Test support
#
# File:   minute_cCodeGenerator.pl
#
# Author: Silvano Catinella <catinella@yahoo.com>
#
# Description:
#	This script generates the code you have to include in your C program's main() function.
#	The generated procedure will manage the program file arguments and run the tests in the proper way considering the
#	specified arguments.
#	
#	To use this script respect the following syntax:
#		minute_cCodeGenerator.pl       \
#			--srcFilename=<filename> \
#			--tgtFilename=<filename> \
#			--template=<filename>    \
#			[--verbose] [--help]
#
#-------------------------------------------------------------------------------------------------------------------------------

use warnings;
use strict;

my $err = 0;
my @testsList;

my $TEMPLATE_DIR = "templates";

my %templDB = (
	"C"   => "cMain.tmpl",
	"CPP" => "cppMain.tmpl"
);

sub helpMsg {
	print("Use: minute_cCodeGenerator.pl --srcFilename=<filename> --tgtFilename=<filename> [--verbose] [--help]\n");
}

sub get_template {
	#
	# Description:
	#	It accept a filename as input argument and returns the template file name used for that type of file
	#
	my $ext = shift;
	my $templ = "";
	$ext =~ s/^.*\.//;
	$ext = uc($ext);
	foreach (keys(%templDB)) {
		if ($_ eq $ext) {
			$templ = $templDB{$_};
			last;
		}
	}
	return($templ);
}
#-------------------------------------------------------------------------------------------------------------------------------
#                                                        M A I N
#-------------------------------------------------------------------------------------------------------------------------------
my $srcfile = "";
my $tgtfile = "";
my $myVerbose = 0;

if (scalar(@ARGV) < 2) {
	# ERROR!
	helpMsg;
	$err = 131;

} else {
	my $myPath = $0;
	$myPath =~ s/\/[^\/]*//;

	{
		my $key;
		my $value;
		foreach (@ARGV) {
			if (/^--[^=]+=/) {
				$key = $_;
				$value = $_;
				$key =~ s/=.*$//;
				$value =~ s/^[^=]+=//;
				if ($key eq "--srcFilename") {
					$srcfile = $value;
				} elsif ($key eq "--tgtFilename") {
					$tgtfile = $value;
				} else {
					# ERROR!
					print(STDERR "ERROR! \"$_\" is unknown file's argument\n");
					helpMsg;
					$err = 133;
				}
			} elsif (/^--[^=]+$/) {
				$key = $_;
				$value = "";
				if ($key eq "--verbose") {
					$myVerbose = 1;
					
				} elsif ($key eq "--help") {
					helpMsg;
					
				} else {
					# ERROR!
					print(STDERR "ERROR! \"$_\" is unknown file's argument\n");
					helpMsg;
					$err = 133;
				}
			}
		}
	}

	if (-f "$srcfile") {
		my $fh;
		my $counter = 0;
		if ($myVerbose) {
			print("Parsing of \"$srcfile\" file...\n");
			print("Recognized tests list:\n");
		}
		
		if (open($fh, "<$srcfile")) {
			while (<$fh>) {
				my %row = ();
				if (/^[ \t]*TEST/) {
					chomp;
					s/[ \t]//g;
					s/^TEST\(([^,]+),([^,]+)\).*/$1__$2/;
					$row{"cbut"}     = $1; # Code Block Under Test
					$row{"ttype"}    = $2; # Test Type
					$row{"testproc"} = $_;
					if ($_) {push(@testsList, \%row)};
					if ($myVerbose) {
						printf("\t%d) %s::%s\n", $counter, $row{"cbut"}, $row{"ttype"});
					}
					$counter++;
				}
			}
			close($fh);
			
			if ($counter == 0) {
				# WARNING!
				print("WARNING! No test found in the \"$srcfile\" file\n");
				$err = 16;
			} else {
				my $template = &get_template($srcfile);

				if ($template eq "") {
					print("ERROR! I cannot find the proper template for the \"$srcfile\" file\n");
					$err = 135;

				} else {
					my $tgtFH;
					my $tmplFH;
					
					if ($myVerbose) {
						print("\nI will use the \"$template\" template file\n");
					}
					
					if (not open($tgtFH, "> $tgtfile")) {
						# ERROR!
						print(STDERR "ERROR! I cannot write the \"$tgtfile\" file\n$!\n");
						$err = 131;

					} elsif (not open($tmplFH, "< $myPath/$TEMPLATE_DIR/$template")) {
						# ERROR!
						print(STDERR "ERROR! I cannot read the \"$template\" file\n$!\n");
						$err = 131;
						close($tgtFH);

					} else {
						my $tNamesList = "";
						my $tFuncsList = "";
						if ($myVerbose) {print("Template \"$template\" compiling:")};
						foreach (@testsList) {
							$tNamesList = $tNamesList . "\"" . $_->{"cbut"} . "::" . $_->{"ttype"} . "\", ";
							$tFuncsList = $tFuncsList . $_->{"testproc"} . ", ";
						}
						$tNamesList =~ s/, *$//;
						$tFuncsList =~ s/, *$//;
						
						print($tgtFH "// This file has been aumatically created by $0 script\n\n");
						while (<$tmplFH>) {
							chomp;
							if (/##[^ \t;]+##/) {
								if ($myVerbose) {print("#")};
								if (/##TESTSNUM##/) {
									s/##TESTSNUM##/$counter/;
								}
								if (/##TESTSLIST##/) {
									s/##TESTSLIST##/$tNamesList/;
								}
								if (/##FUNCSLIST##/) {
									s/##FUNCSLIST##/$tFuncsList/;
								}
							} else {
								if ($myVerbose) {print(".")};
							}
							print($tgtFH "$_\n");
						}
						if ($myVerbose) {print("\n")};
						close($tgtFH);
						close(tmplFH);
					}
				}
			}
		} else {
			# ERROR!
			print(STDERR "ERROR! I cannot read the \"$srcfile\" file\n");
			$err = 131;
		}
	} else {
		# ERROR!
		print(STDERR "ERROR! \"$srcfile\" file not found\n");
		$err = 129;
	}	
}

exit($err);

File: minute_cCodeGenerator.py
#!/usr/bin/python3
#-------------------------------------------------------------------------------------------------------------------------------
#                                                                                                       
#                            _/      _/  _/                        _/    _/          _/_/_/_/_/         
#                           _/_/  _/_/      _/_/_/                _/    _/              _/      _/_/    
#                          _/  _/  _/  _/  _/    _/  _/_/_/_/_/  _/    _/  _/_/_/_/_/  _/    _/_/_/_/   
#                         _/      _/  _/  _/    _/              _/    _/              _/    _/          
#                        _/      _/  _/  _/    _/                _/_/                _/      _/_/_/     
#                                                                                                       
#                                                                                                       
#                                                    Minimal Unit Test support
#
# File:   minute_cCodeGenerator.py
#
# Author: Silvano Catinella <catinella@yahoo.com>
#
# Description:
#	This script generates the code you have to include in your C program's main() function.
#	The generated procedure will manage the program file arguments and run the tests in the proper way considering the
#	specified arguments.
#	
#	To use this script respect the following syntax:
#		minute_cCodeGenerator.pl       \
#			--srcFilename=<filename> \
#			--tgtFilename=<filename> \
#			--template=<filename>    \
#			[--verbose] [--help]
#
#-------------------------------------------------------------------------------------------------------------------------------+
import sys
import argparse
import re
from pathlib import Path

TEMPLATE_DIR = "templates";
TESTSNUM_PATTERN = r"##TESTSNUM##"
TESTSLIST_PATTERN = r"##TESTSLIST##"
FUNCSLIST_PATTERN = r"##FUNCSLIST##"

templDB = {
	"c"   : "cMain.tmpl",
	"cpp" : "cppMain.tmpl"
}

def testsFileParser (filename: str, verbose: bool) -> list[str] :
	#
	# Description:
	#	It parses the src-file that contain TEST macros ansd returns a list of the defined tests names
	#
	fh = open(filename, "r");
	pattern = r"TEST[\t]*(.+,.t)"
	tlist = []
	if (verbose) :
		print("\nDetected test procedures:")
	while (row := fh.readline()) :
		row = row.rstrip("\n")
		if (re.match(pattern, row)):
			row = re.sub(r"[ \t]", r"", row);
			item = re.sub(r"TEST\(([^,]+),([^,]+)\).*", r"\1::\2", row);
			if (verbose) :
				print(f"\t{item}");
			tlist.append(item);
	fh.close()
	return(tlist)

def get_template (filename: str) -> str :
	#
	# Description:
	#	It accept a filename as input argument and returns the template file name used for that type of file
	#
	templFile = ""
	ext = re.sub(r"^.*\.", r"", filename);
	for key in templDB.keys() :
		if (key.upper() == ext.upper()) :
			templFile = templDB[key]

	return(templFile)

#-------------------------------------------------------------------------------------------------------------------------------
#                                                        M A I N
#-------------------------------------------------------------------------------------------------------------------------------
parser = argparse.ArgumentParser()

parser.add_argument(
	"--srcFilename",
	required = True,
	action   = "store",
	help     = "The source file that contains the tests"
)
parser.add_argument(
	"--tgtFilename",
	required = True,
	action   = "store",
	help     = "The name of the file that contains the main() function definition"
)
parser.add_argument(
	"--verbose",
	action = "store_true",
	help   = "It increcmets the process' output verbosity level"
)

args = parser.parse_args()

#
# Checking for the file's arguments
#
if (Path(args.srcFilename).is_file()) == False :
	print(f"ERROR! \"{args.srcFilename}\" file not found")
	exit(129)

else :
	tNFList  = ""
	tFFList  = ""
	myPath   = re.sub(r"[^\/]*$", "", sys.argv[0])
	testList = testsFileParser(args.srcFilename, args.verbose)
	tmplFile = get_template(args.srcFilename)
	
	for rec in testList :
		tNFList = tNFList + f'"{rec}"' + ", "
		rec = re.sub(r"::", r"__", rec)
		tFFList = tFFList + rec + ", ";
		
	tNFList = re.sub(r", $", "", tNFList)
	tFFList = re.sub(r", $", "", tFFList)
						
	if args.verbose :
		print(f"Selected template: {tmplFile}");
	
	try :
		tmplFh = open(f"{myPath}/{TEMPLATE_DIR}/{tmplFile}", "r")
		sgcFh  = open(args.tgtFilename, "w")

	except OSError as err:
		print(err)
		exit(131)
	
	while (row := tmplFh.readline()) :
		#row = row.rstrip("\n")
		if (re.match(r".*##[^#]+##", row)) :
			row = re.sub(TESTSNUM_PATTERN,  str(len(testList)), row)
			row = re.sub(TESTSLIST_PATTERN, tNFList, row)
			row = re.sub(FUNCSLIST_PATTERN, tFFList, row)
		
		sgcFh.write(row)
	
	tmplFh.close();
	sgcFh.close();

